src/ui/colorLegendWidget.js

```diff
@@ -0,0 +1,421 @@
+  this._renderDiscreteColors = function (svg, steps, colorScale, width, getValue) {
+    svg.selectAll('rect')
+      .data(steps)
+      .enter()
+      .append('rect')
+      .attr('width', width / steps.length)
+      .attr('height', '20px')
+      .attr('fill', function (d) {
+        return colorScale(d);
+      })
+      .attr('transform', function (d, i) {
+        return 'translate(' + i * width / steps.length + ' ,0)';
+      })
+      .on('mousemove', function (d) {
+        m_this._showPopup(getValue(d));
+      })
+      .on('mouseout', m_this._hidePopup);
+  };
+
+  /**
+   * Draw an individual continous type legend.
+   * @param {Element} svg svg element that the legend will be drawn
+   * @param {number} width width of the svg element in pixel
+   * @param {geo.gui.colorLegendWidget.category} category The continuous type legend category
+   */
+  this._drawContinous = function (svg, width, category) {
+    var axisScale, axis;
+    if (['linear', 'log', 'sqrt', 'pow'].indexOf(category.scale) === -1) {
+      throw new Error('unsupported scale');
+    }
+    axisScale = d3.scale[category.scale]().domain(category.domain).range([0, width]).nice();
+    if (category.scale === 'log' && category.base) {
+      axisScale.base(category.base);
+    }
+    if (category.scale === 'pow' && category.exponent) {
+      axisScale.exponent(category.exponent);
+    }
+    var id = uniqueID();
+    var precision = Math.max.apply(null, category.domain
+      .map(function (number) { return getPrecision(number); }));
+
+    var gradient = svg
+      .append('defs')
+      .append('linearGradient')
+      .attr('id', 'gradient' + id);
+    gradient.append('stop')
+      .attr('offset', '0%')
+      .attr('stop-color', category.colors[0]);
+    gradient.append('stop')
+      .attr('offset', '100%')
+      .attr('stop-color', category.colors[1]);
+    svg.append('rect')
+      .attr('fill', 'url(#gradient' + id + ')')
+      .attr('width', width)
+      .attr('height', '20px')
+      .on('mousemove', function () {
+        var value = axisScale.invert(d3.mouse(this)[0]);
+        var text = m_this._popupFormatter(value, precision);
+        m_this._showPopup(text);
+      })
+      .on('mouseout', m_this._hidePopup);
+
+    axis = d3.svg.axis()
+      .scale(axisScale)
+      .ticks(m_ticks, '.2s');
+
+    this._renderAxis(svg, axis);
+  };
+
+  /**
+   * Actually render the axis with d3.
+   * @param {Element} svg svg element that the axis will be drawn
+   * @param {object} axis d3 axis object
+   */
+  this._renderAxis = function (svg, axis) {
+    svg.append('g')
+      .attr('class', 'axis x')
+      .attr('transform', 'translate(0, 20)')
+      .call(function (g) {
+        g.call(axis);
+      });
+  };
+
+  /**
+   * Formatter of number that tries to maximize the precision
+   * while making the output shorter.
+   * @param {number} number to be formatted
+   * @param {precision} precision number of decimal precision will be tried to be kept
```

Comment:
I think the data type should just be `number`.  I'd recommend changing this to `@param {number} precision Maximum number of decimal places that are kept.`.